Src/FluentAssertions/Specialized/GenericAsyncFunctionAssertions.cs

```diff
@@ -51,7 +51,7 @@ public GenericAsyncFunctionAssertions(Func<Task<TResult>> subject, IExtractExcep
 
             if (success)
             {
-                bool completesWithinTimeout = await CompletesWithinTimeoutAsync(task, remainingTime);
+                bool completesWithinTimeout = await CompletesWithinTimeoutAsync(task, remainingTime, _ => Task.CompletedTask);
```

Comment:
On a second thought, if we start returning `default` on canceled generic tasks instead of propagating the cancellation exception, we could cause the confusion that `default` was the _successful_ value.

Reply:
I'm sorry, but I'm completely at loss of what you are trying to say here. Are you suggesting a different fix? If so, please take over this PR and finish it as you would like to see it.

Reply:
Sorry for the lengthy brain dump, I got confused myself staring at this problem.

I've added some tests to show the inconsistency we will now have for `CompleteWithinAsync` between `Task` and `Task<T>`.
These added tests have to fail on cancellation as we cannot distinguish between:
* FA treating cancellation as success and returning `default`
  * https://github.com/fluentassertions/fluentassertions/blob/e0f45049d8acadae0c98f9710b650dd413480355/Src/FluentAssertions/Specialized/GenericAsyncFunctionAssertions.cs#L63
* The subject task running to completion and returning `default`

I don't like the inconsistency between `Task` and `Task<T>` but I prefer that over a false positives. 

The only way I see to "fix" this would be to change the semantics of `CompleteWithinAsync` 
from "will complete within the specified time" 
to "will run to completion within the specified time".

Reply:
> The only way I see to "fix" this would be to change the semantics of `CompleteWithinAsync`
> from "will complete within the specified time"
> to "will run to completion within the specified time".

And with that you mean that a canceled task is also included in the second case, but not in the first?

Reply:
Not sure what you mean by "included" 😞 

With the risk of repeating myself I'll try with examples instead 🤞

With this PR `CompleteWithinAsync` passes when a `Task` is canceled.
```cs
[Fact]
public async Task Non_generic_task()
{
    // Arrange
    var timer = new FakeClock();
    var taskFactory = new TaskCompletionSource(); // <-- non-generic

    // Act
    Func<Task> action = () => taskFactory
        .Awaiting(t => t.Task)
        .Should(timer)
        .CompleteWithinAsync(100.Milliseconds()); // <-- AsyncFunctionAssertions

    taskFactory.SetCanceled();
    timer.Complete();

    // Assert
    await action.Should().NotThrowAsync(); // <-- passes
}
```

On the other hand `CompleteWithinAsync` for `Task<T>` does _not_ treat cancellation as success.
```cs
[Fact]
public async Task Generic_task_canceled()
{
    // Arrange
    var timer = new FakeClock();
    var taskFactory = new TaskCompletionSource<bool>(); // <-- generic

    // Act
    Func<Task> action = () => taskFactory
        .Awaiting(t => t.Task)
        .Should(timer)
        .CompleteWithinAsync(100.Milliseconds()) // <-- GenericAsyncFunctionAssertions
        .WithResult(false);

    taskFactory.SetCanceled();
    timer.Complete();

    // Assert
    await action.Should().ThrowAsync<TaskCanceledException>(); // <-- to avoid false-positive between a canceled task and subject Task<T> returning default(bool).
}
```

The reason be that when the subject `Task<T>` is canceled, we cannot pass along `default(T)` for further asserting, as that would conflict with a subject `Task<T>` successfully returning `default(T)`.
```cs
[Fact]
public async Task Generic_task_returning_default()
{
    // Arrange
    var timer = new FakeClock();
    var taskFactory = new TaskCompletionSource<bool>(); // <-- generic

    // Act
    Func<Task> action = () => taskFactory
        .Awaiting(t => t.Task)
        .Should(timer)
        .CompleteWithinAsync(100.Milliseconds()) // <-- GenericAsyncFunctionAssertions
        .WithResult(false);

    taskFactory.SetResult(false); // <-- the subject Task<bool> returns default(bool)
    timer.Complete();

    // Assert
    await action.Should().NotThrowAsync(); // <-- passes
}
```

Reply:
Ah, I think I finally understand.